### PR TITLE
Wait for clusterState changes before invoking retries

### PR DESCRIPTION
**Pull request by mfussenegger**
This commit makes sure that we wait for the clusterState to have changed
before invoking a retry. This should reduce the chance for the retry to
fail again, which is likely the case if it used the same cluster state
again to calculate the routing.

### CODE REVIEW

**Comment by mfussenegger** [comment]
retest this please

File: blackbox/docs/src/doc_tests/process_test.py [hunk 1]
```diff
@@ -380,10 +380,5 @@ def exec_select_queries(client, is_active, errors, finished):
             try:
                 client.sql('select name, count(*) from t1 group by name')
             except Exception as e:
-                # For now we make sure that queries are not killed, but a
-                # TableUnknownException can happen if the queries still target
-                # the node that is being decommissioned and the cluster state
-                # update is too slow (and the internal 3 retries all happen before)
-                if 'TableUnknownException' not in str(e):
```
Comment (mfussenegger) [inline blackbox/docs/src/doc_tests/process_test.py]: I'm not sure if this is now really fixed. I ran it for quite a while but as it is with race conditions one can never be sure.

File: sql/src/main/java/io/crate/protocols/postgres/RetryOnFailureResultReceiver.java [hunk 2]
```diff
@@ -73,14 +88,36 @@ public void allFinished(boolean interrupted) {
...
+                clusterStateObserver.waitForNextChange(new ClusterStateObserver.Listener() {
+                    @Override
+                    public void onNewClusterState(ClusterState state) {
+                        attempt += 1;
+                        retry();
+                    }
+
+                    @Override
+                    public void onClusterServiceClose() {
+                        delegate.fail(error);
```
Comment (matriv) [inline sql/src/main/java/io/crate/protocols/postgres/RetryOnFailureResultReceiver.java]: Is it ok that we call `delegate.fail(error)` with the unwrapped error now?
Why was this change needed?

Comment (mfussenegger) [reply sql/src/main/java/io/crate/protocols/postgres/RetryOnFailureResultReceiver.java]: We already used the unwrapped error before. (`t = SQLExceptions.unwrap(t)`. I had to introduce a separate variable because it's now used in the anonymous Listener class which requires an effectively final variable.

Comment (matriv) [reply sql/src/main/java/io/crate/protocols/postgres/RetryOnFailureResultReceiver.java]: ok,Thx! somehow missed that... :-(
